lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -311,3 +312,83 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Initialize a ConcreteExistentialInfo based on the already computed concrete
+/// type and protocol declaration. It first determines the OpenedArchetypeDef
+/// from the Arg argument and then appends new unchecked_cast instructions in
+/// the function body using the SILBuilder.
+ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &ArgOperand,
+                                                 const CanType &ConcreteTy,
+                                                 ProtocolDecl *Protocol,
+                                                 SILBuilder &Builder)
+    : ConcreteType(ConcreteTy) {
+  SILValue Arg = ArgOperand.get();
+  SILInstruction *User = ArgOperand.getUser();
+
+  // We need the OpenedArchetypeDef to be able to cast it to the concrete type.
+  // findInitExistential helps us determine this OpenedArchetypeDef. For cases
+  // where OpenedArchetypeDef can not be found from findInitExistential (because
+  // there was no InitExistential), then the code after findInitExistential does
+  // extra effort in trying to find an OpenedArchetypeDef for AllocStackInst
+  // using getAddressOfStackInit.
+  isCopied = false;
+  OpenedArchetypeDef = Arg;
+  InitExistential = findInitExistential(ArgOperand, OpenedArchetype,
+                                        OpenedArchetypeDef, isCopied);
```

Comment:
Thanks for the comments. If InitExistential is non-null, do you think it would be useful to assert that it's concrete type is the same as the passed in ConcreteType?

Reply:
Good idea!